numba/passthru.py

```diff
@@ -0,0 +1,241 @@
+        ]
+        super(PassThruModel, self).__init__(dmm, fe_typ, members)
+
+
+@unbox(PassThruType)
+def unbox_pass_thru_type(typ, obj, context):
+    pass_thru = cgutils.create_struct_proxy(typ)(
+        context.context, context.builder
+    )
+    pass_thru.meminfo = context.pyapi.nrt_meminfo_new_from_pyobject(obj, obj)
+
+    return NativeValue(pass_thru._getvalue())
+
+
+@box(PassThruType)
+def box_pass_thru_type(typ, val, context):
+    val = cgutils.create_struct_proxy(typ)(
+        context.context, context.builder, value=val
+    )
+    obj = context.context.nrt.meminfo_data(context.builder, val.meminfo)
+
+    context.pyapi.incref(obj)
+    context.context.nrt.decref(context.builder, typ, val._getvalue())
+
+    return obj
+
+
+class PassThruContainer(object):
+    """A container to ferry arbitrary Python objects through *nopython* mode.
+       The only operation supported in *nopython-mode* is ``==``. Two instances
+       of ``PassThruContainer`` are equal if the wrapped objects are identical,
+       ie if ``a.obj is b.obj``.
+    """
+    def __init__(self, obj):
+        self._obj = obj
+
+    @property
+    def obj(self):
+        return self._obj
+
+    def __eq__(self, other):
+        return isinstance(other, PassThruContainer) and self.obj is other.obj
+
+    def __hash__(self):
+        return object.__hash__(self.obj)
+
+
+class PassThruContainerType(PassThruType):
+    def __init__(self):
+        super(PassThruContainerType, self).__init__()
+
+
+pass_thru_container_type = PassThruContainerType()
+
+
+@typeof_impl.register(PassThruContainer)
+def type_pass_thru_container(val, context):
+    return pass_thru_container_type
+
+
+@register_model(PassThruContainerType)
+class PassThruContainerModel(models.StructModel):
+    def __init__(self, dmm, fe_typ):
+        members = [
+            ('container', pass_thru_type),
+            ('wrapped_obj', opaque_pyobject)
+        ]
+        super(PassThruContainerModel, self).__init__(dmm, fe_typ, members)
+
+
+make_attribute_wrapper(PassThruContainerType, 'wrapped_obj', 'wrapped_obj')
+
+
+@unbox(PassThruContainerType)
+def unbox_pass_thru_container_type(typ, obj, context):
+    container = cgutils.create_struct_proxy(typ)(
+        context.context, context.builder
+    )
+
+    container.container = context.unbox(pass_thru_type, obj).value
+
+    wrapped_obj = context.pyapi.object_getattr_string(obj, "obj")
+    context.pyapi.decref(wrapped_obj)
+    container.wrapped_obj = wrapped_obj
+
+    return NativeValue(container._getvalue())
+
+
+@box(PassThruContainerType)
+def box_pass_thru_container_type(typ, val, context):
+    val = cgutils.create_struct_proxy(typ)(
+        context.context, context.builder, value=val
+    )
+
+    return context.box(pass_thru_type, val.container)
+
+
+@lower_builtin(int, types.Opaque)
+def opaque_to_int(context, builder, sig, args):
+    return builder.ptrtoint(args[0], cgutils.intp_t)
```

Comment:
Wouldn't this make more sense as an overload of `id`?

Reply:
Was not sure whether to name it `int` or `id` myself. Guess I'll rename to `id` then.

Reply:
If you call it `id`, are the semantics of `id(pass_through_container.obj)` the same inside and outside of jitted code? If yes, then `id` is the right name.

Reply:
This probably then belongs in `cpython/builtins.py`.